### PR TITLE
Feat(web, web-twig, web-react): Hide close button in Modal

### PR DESCRIPTION
**Pull request by pavelklibani**
## Description

- Added a new prop to hide the close button in the `ModalHeader`.
- Added a new prop to disable the escape key from closing the modal.

### Additional context

TODO:
- [x] Update visual tests
- [x] Update storybook

### Issue reference

[Modal: komponenta nemusí mít zavírací tlačítko](https://jira.almacareer.tech/browse/DS-1063)

### CODE REVIEW

File: packages/web-react/src/components/Dialog/Dialog.tsx [hunk 1]
```diff
@@ -7,7 +7,7 @@ import { useDialog } from './useDialog';
 // Solved using `as MutableRefObject<HTMLDialogElement | null>` but I do not like it
 
 const Dialog = (props: DialogProps, ref: ForwardedRef<HTMLDialogElement | null>): JSX.Element => {
-  const { children, isOpen, onClose, closeOnBackdropClick = true, ...restProps } = props;
+  const { children, isOpen, onClose, closeOnBackdropClick = true, closeOnEscapeKeyDown, ...restProps } = props;
```
Comment (literat) [inline packages/web-react/src/components/Dialog/Dialog.tsx:10]:
```suggestion
  const { children, isOpen, onClose, closeOnBackdropClick = true, closeOnEscapeKeyDown = true, ...restProps } = props;
```

Maybe if you use the default value, you shouldn't set the type by `as boolean` 🤔

File: packages/web-react/src/components/Modal/README.md [hunk 2]
```diff
@@ -165,12 +166,26 @@ accessible name for the dialog, e.g. using the `aria-label` attribute on
...
 
-| Name         | Type        | Default | Required | Description        |
-| ------------ | ----------- | ------- | -------- | ------------------ |
-| `children`   | `ReactNode` | —       | ✕        | Children node      |
-| `closeLabel` | `string`    | —       | ✕        | Close button label |
+| Name              | Type        | Default | Required | Description                    |
+| ----------------- | ----------- | ------- | -------- | ------------------------------ |
+| `children`        | `ReactNode` | —       | ✕        | Children node                  |
+| `closeLabel`      | `string`    | `Close` | ✕        | Close button label             |
+| `hideCloseButton` | `bool`      | `false` | ✕        | Whether close button is hidden |
```
Comment (literat) [inline packages/web-react/src/components/Modal/README.md:188]:
```suggestion
| `hideCloseButton` | `bool`      | `false` | ✕        | Whether close button is rendered |
```

I think we do not just hide the butter. The button is not rendered, I think.

File: packages/web-react/src/components/Modal/demo/ModalHiddenCloseButton.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import React, { useState } from 'react';
+import { Button } from '../../Button';
+import Modal from '../Modal';
+import ModalBody from '../ModalBody';
+import ModalDialog from '../ModalDialog';
+import ModalFooter from '../ModalFooter';
+import ModalHeader from '../ModalHeader';
+
+const ModalHiddenCloseButton = () => {
+  const [isOpen, setOpen] = useState(false);
```
Comment (literat) [inline packages/web-react/src/components/Modal/demo/ModalHiddenCloseButton.tsx:10]: Btw. we have `useToggle` hook with the same functionality as these three lines ;-)

File: packages/web-react/src/components/Dialog/Dialog.tsx (hunk 1 above)
Comment (pavelklibani) [reply packages/web-react/src/components/Dialog/Dialog.tsx:10]: I wasn't sure where to set this default, so I decided to set it inside the `useCancelEvent` hook and not here, because it is just passing through this component.

File: packages/web-react/src/hooks/useCancelEvent.ts [hunk 4]
```diff
@@ -1,25 +1,61 @@
...
 
-      callback(event);
+      if (callback && closeOnEscapeKeyDown) {
+        callback(event);
+      }
+    },
+    [ref, callback, closeOnEscapeKeyDown],
+  );
+
+  // ! Removed when bug in chrome will be fixed
```
Comment (literat) [inline packages/web-react/src/hooks/useCancelEvent.ts:39]:
```suggestion
  // ! Removed when bug in Chrome will be fixed
```

File: packages/web-react/src/hooks/useCancelEvent.ts [hunk 5]
```diff
@@ -1,25 +1,61 @@
-import { useCallback, useEffect, MutableRefObject } from 'react';
+import { useCallback, MutableRefObject } from 'react';
+import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
+
+// TODO: Remove `isOpen` and listeners with `handleKeyDown` when Chrome fixes the bug,
```
Comment (literat) [inline packages/web-react/src/hooks/useCancelEvent.ts:4]: Do you find any bug reports related to this issue? Maybe some links to it will be helpful.

File: packages/web-react/src/hooks/useCancelEvent.ts [hunk 6]
```diff
@@ -1,25 +1,61 @@
...
+        callback(event);
+      }
+    },
+    [ref, callback, closeOnEscapeKeyDown],
+  );
+
+  // ! Removed when bug in chrome will be fixed
+  const handleKeyDown = useCallback(
+    (event: KeyboardEvent) => {
+      if (event.key === 'Escape' && !closeOnEscapeKeyDown && isOpen) {
```
Comment (literat) [inline packages/web-react/src/hooks/useCancelEvent.ts:42]:
```suggestion
      if (event.key === KEY_ESCAPE && !closeOnEscapeKeyDown && isOpen) {
```

Please, set a constant for this :-)

File: packages/web/src/js/Modal.ts [hunk 7]
```diff
@@ -53,13 +57,31 @@ class Modal extends BaseComponent {
...
+
+    if (closeOnEscapeKeyDown === false) {
+      event.preventDefault();
+    }
+  }
+
+  handleKeyDown(event: KeyboardEvent) {
+    const { closeOnEscapeKeyDown } = this.config as Config;
+
+    if (event.key === 'Escape' && closeOnEscapeKeyDown === false && this.isShown) {
```
Comment (literat) [inline packages/web/src/js/Modal.ts:71]: Please, set a const for `Escape` string.

File: packages/web-react/src/components/Modal/demo/index.tsx [hunk 8]
```diff
@@ -27,6 +28,9 @@ ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
       <DocsSection title="Disabled Backdrop Click">
         <ModalDisabledBackdropClick />
       </DocsSection>
+      <DocsSection title="Hidden Close Button">
```
Comment (literat) [inline packages/web-react/src/components/Modal/demo/index.tsx:31]: It is not just a hidden close button, it is also a disabled escape key. Maybe something like `Dialog choice` or similar naming for simple actions like yes/no, confirm/cancel.

File: packages/web-react/src/components/Dialog/Dialog.tsx (hunk 1 above)
Comment (literat) [reply packages/web-react/src/components/Dialog/Dialog.tsx:10]: What about updating the `DialogProps` type?

Comment (literat) [reply packages/web-react/src/components/Dialog/Dialog.tsx:10]:
```diff
+  closeOnBackdropClick?: boolean;
+  closeOnEscapeKeyDown?: boolean;
```

**Comment by literat** [comment]
Great Job 👍 It is working now. I just wonder if there is really some issue with the Chrome Browser and we should report it or if this is a problem with our logic. Since this was working in Safari and Firefox with just `event.preventDefault` maybe it should be reported by us :-)

File: packages/web-react/src/components/Dialog/Dialog.tsx (hunk 1 above)
Comment (pavelklibani) [reply packages/web-react/src/components/Dialog/Dialog.tsx:10]: types updated

File: packages/web-react/src/components/Modal/README.md (hunk 2 above)
Comment (pavelklibani) [reply packages/web-react/src/components/Modal/README.md:188]: fixed

File: packages/web-react/src/components/Modal/demo/ModalHiddenCloseButton.tsx (hunk 3 above)
Comment (pavelklibani) [reply packages/web-react/src/components/Modal/demo/ModalHiddenCloseButton.tsx:10]: changed

File: packages/web-react/src/hooks/useCancelEvent.ts (hunk 4 above)
Comment (pavelklibani) [reply packages/web-react/src/hooks/useCancelEvent.ts:39]: fixed

File: packages/web-react/src/hooks/useCancelEvent.ts (hunk 6 above)
Comment (pavelklibani) [reply packages/web-react/src/hooks/useCancelEvent.ts:42]: constant created

File: packages/web-react/src/components/Modal/README.md [hunk 9]
```diff
@@ -165,12 +166,26 @@ accessible name for the dialog, e.g. using the `aria-label` attribute on
 </Modal>
 ```
 
+### Hidden Close Button
+
+The close button can by hidden by adding `hideCloseButton` prop to `ModalHeader` component.
```
Comment (dlouhak) [inline packages/web-react/src/components/Modal/README.md:171]:
```suggestion
The close button can be hidden by adding the `hideCloseButton` prop to the `ModalHeader` component.
```

File: packages/web-react/src/components/Modal/ModalHeader.tsx [hunk 10]
```diff
@@ -6,8 +6,13 @@ import ModalCloseButton from './ModalCloseButton';
 import { useModalContext } from './ModalContext';
 import { useModalStyleProps } from './useModalStyleProps';
 
+const defaultProps: ModalHeaderProps = {
+  hideCloseButton: false,
```
Comment (dlouhak) [inline packages/web-react/src/components/Modal/ModalHeader.tsx:10]: Hmm, I have to say I'm a little bit confused when you use prefix [`has`/`is`](https://github.com/lmc-eu/spirit-design-system/pull/1533#discussion_r1664003667) and where not.

Comment (literat) [reply packages/web-react/src/components/Modal/ModalHeader.tsx:10]: 🤔 When the prop is of type Boolean the `is/has` prop should be used to indicate this and thus be more readable. In some cases, we also use `enable` prefix for some options. There exists `hideOnCollapse` on the Accordion component also. In this case, I am not sure if we want to create the easy-to-use API `<ModalHeader hideCloseButton />` or something worse usable to suggest that this is not a good idea like `<ModalHeader hasCloseButton={false} />`.

**Review by literat — APPROVED** [review]
LGTM 👍  Please, solve with the others the prop names. 🙏

File: packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx [hunk 11]
```diff
@@ -10,4 +12,24 @@ describe('ModalHeader', () => {
   stylePropsTest(ModalHeader);
 
   restPropsTest(ModalHeader, 'header');
+
+  it('should have close button', () => {
+    render(<ModalHeader id="test">Modal Title</ModalHeader>);
+
+    const closeButton = screen.getByRole('button');
+
+    expect(closeButton).toBeInTheDocument();
```
Comment (curdaj) [inline packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx:21]:
```suggestion
    expect(screen.getByRole('button')).toBeInTheDocument();
```

File: packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx [hunk 12]
```diff
@@ -10,4 +12,24 @@ describe('ModalHeader', () => {
   stylePropsTest(ModalHeader);
 
   restPropsTest(ModalHeader, 'header');
+
+  it('should have close button', () => {
+    render(<ModalHeader id="test">Modal Title</ModalHeader>);
+
+    const closeButton = screen.getByRole('button');
```
Comment (curdaj) [inline packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx:19]:
```suggestion

```

File: packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx [hunk 13]
```diff
@@ -10,4 +12,24 @@ describe('ModalHeader', () => {
...
+  });
+
+  it('should not have close button', () => {
+    render(
+      <ModalHeader id="test" hideCloseButton>
+        Modal Title
+      </ModalHeader>,
+    );
+
+    const closeButton = screen.queryByRole('button');
```
Comment (curdaj) [inline packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx:31]:
```suggestion

```

File: packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx [hunk 14]
```diff
@@ -10,4 +12,24 @@ describe('ModalHeader', () => {
...
+
+    expect(closeButton).not.toBeInTheDocument();
```
Comment (curdaj) [inline packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx:33]:
```suggestion
    expect(screen.getByRole('button')).not.toBeInTheDocument();
```

**Review by curdaj — APPROVED** [review]

File: packages/web-react/src/components/Modal/ModalHeader.tsx (hunk 10 above)
Comment (dlouhak) [reply packages/web-react/src/components/Modal/ModalHeader.tsx]: Discussed with @crishpeen: `hasCloseButton` is good enough but we should talk about this topic during Great (Long) Topics in the future.

File: packages/web-react/src/components/Modal/README.md [hunk 15]
```diff
@@ -165,12 +166,26 @@ accessible name for the dialog, e.g. using the `aria-label` attribute on
 </Modal>
 ```
 
+### Hidden Close Button
+
+The close button can be hidden by adding the `hideCloseButton` prop to the `ModalHeader` component.
```
[the remark this answers is not in the document]
Comment (crishpeen) [reply packages/web-react/src/components/Modal/README.md:171]:
```suggestion
The close button can be hidden by adding the `hasCloseButton` prop to the `ModalHeader` component.
```

[the remark this answers is not in the document]
Comment (crishpeen) [reply packages/web-react/src/components/Modal/README.md:171]: And maybe reword it so it matches the prop? We are not hiding it we remove it.

```suggestion
To render the `ModalHeader` component without the close button set the `hasCloseButton` prop to false.
```

File: packages/web-react/src/components/Modal/ModalHeader.tsx (hunk 10 above)
Comment (pavelklibani) [reply packages/web-react/src/components/Modal/ModalHeader.tsx]: prop renamed to to `hasCloseButton` 👍

File: packages/web-react/src/components/Modal/demo/ModalHiddenCloseButton.tsx [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+      <Modal
+        id="example-hidden-close-button"
+        isOpen={isOpen}
+        onClose={handleClose}
+        closeOnBackdropClick={false}
+        closeOnEscapeKeyDown={false}
+      >
+        <ModalDialog>
+          <ModalHeader id="example-hidden-close-button" hasCloseButton={false}>
```
Comment (crishpeen) [inline packages/web-react/src/components/Modal/demo/ModalHiddenCloseButton.tsx:26]: do we need the id? If so, please make it unique

File: packages/web-react/src/types/modal.ts [hunk 17]
```diff
@@ -17,11 +17,12 @@ export type ModalDialogHandlingProps = {
   isOpen: boolean;
   onClose: (event: ClickEvent) => void;
   closeOnBackdropClick?: boolean;
+  closeOnEscapeKeyDown?: boolean;
 };
 
 export interface ModalCloseButtonProps extends ModalDialogHandlingProps {
   id: string;
-  label: string;
+  label?: string;
```
Comment (crishpeen) [inline packages/web-react/src/types/modal.ts:25]: is it actually optional? I don't think so

**Review by curdaj — APPROVED** [review]

File: packages/web-twig/src/Resources/components/Modal/README.md [hunk 18]
```diff
@@ -158,14 +159,30 @@ using the `aria-label` attribute on `<Modal>` component:
 </Modal>
 ```
 
+### Hidden Close Button
+
+The close button can by hidden by adding `hasCloseButton={false}` to `ModalHeader` component.
```
Comment (crishpeen) [inline packages/web-twig/src/Resources/components/Modal/README.md:164]: please update this as in react

File: packages/web-twig/src/Resources/components/Modal/README.md [hunk 19]
```diff
@@ -158,14 +159,30 @@ using the `aria-label` attribute on `<Modal>` component:
...
-| --------------- | -------- | ------- | -------- | ----------------------- |
-| `closeLabel`    | `string` | `Close` | ✕        | Custom close label      |
-| `enableDismiss` | `bool`   | `true`  | ✕        | Enable JS Modal dismiss |
-| `modalId`       | `string` | —       | ✓        | Modal ID                |
-| `titleId`       | `string` | `null`  | ✕        | ID of the title         |
+| Name             | Type     | Default | Required | Description                    |
+| ---------------- | -------- | ------- | -------- | ------------------------------ |
+| `closeLabel`     | `string` | `Close` | ✕        | Custom close label             |
+| `enableDismiss`  | `bool`   | `true`  | ✕        | Enable JS Modal dismiss        |
+| `hasCloseButton` | `bool`   | `true`  | ✕        | Whether close button is hidden |
```
Comment (crishpeen) [inline packages/web-twig/src/Resources/components/Modal/README.md:183]: rendered? As in react

File: packages/web/src/js/Modal.ts [hunk 20]
```diff
@@ -53,13 +58,31 @@ class Modal extends BaseComponent {
...
+
   addEventListeners() {
     EventHandler.on(this.element, 'close', (event: KeyboardEvent) => this.onDialogClose(event));
     if (this.isTouchDevice) {
       EventHandler.on(window, 'touchstart', (event: Event & { target: Window }) => this.onClick(event));
     } else {
       EventHandler.on(window, 'click', (event: Event & { target: Window }) => this.onClick(event));
     }
+    EventHandler.on(this.element, 'cancel', (event: Event) => this.handleCancel(event));
+    EventHandler.on(document, 'keydown', (event: KeyboardEvent) => this.handleKeyDown(event));
```
Comment (crishpeen) [inline packages/web/src/js/Modal.ts:85]: shouldn't we mention the Chrome bug here (in this file) as well?

Comment (pavelklibani) [reply packages/web/src/js/Modal.ts:85]: comment added

File: packages/web-twig/src/Resources/components/Modal/README.md (hunk 19 above)
Comment (pavelklibani) [reply packages/web-twig/src/Resources/components/Modal/README.md:183]: fixed

File: packages/web-twig/src/Resources/components/Modal/README.md (hunk 18 above)
Comment (pavelklibani) [reply packages/web-twig/src/Resources/components/Modal/README.md:164]: updated

File: packages/web-react/src/hooks/useCancelEvent.ts [hunk 21]
```diff
@@ -1,25 +1,62 @@
...
 
-      callback(event);
+      if (callback && closeOnEscapeKeyDown) {
+        callback(event);
+      }
+    },
+    [ref, callback, closeOnEscapeKeyDown],
+  );
+
+  // ! Removed when bug in Chrome will be fixed
```
[the remark this answers is not in the document]
Comment (pavelklibani) [reply packages/web-react/src/hooks/useCancelEvent.ts:40]: i removed all this comments, just left the main on the top

File: packages/web-react/src/components/Modal/demo/ModalHiddenCloseButton.tsx (hunk 16 above)
Comment (pavelklibani) [reply packages/web-react/src/components/Modal/demo/ModalHiddenCloseButton.tsx:26]: we don't, id removed

File: packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx [hunk 22]
```diff
@@ -10,4 +12,20 @@ describe('ModalHeader', () => {
...
+
+  it('should have close button', () => {
+    render(<ModalHeader id="test">Modal Title</ModalHeader>);
+
+    expect(screen.getByRole('button')).toBeInTheDocument();
+  });
+
+  it('should not have close button', () => {
+    render(
+      <ModalHeader id="test" hasCloseButton={false}>
```
[the remark this answers is not in the document]
Comment (pavelklibani) [reply packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx:24]: we don't, id removed

File: packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx [hunk 23]
```diff
@@ -10,4 +12,20 @@ describe('ModalHeader', () => {
   stylePropsTest(ModalHeader);
 
   restPropsTest(ModalHeader, 'header');
+
+  it('should have close button', () => {
+    render(<ModalHeader id="test">Modal Title</ModalHeader>);
```
[the remark this answers is not in the document]
Comment (pavelklibani) [reply packages/web-react/src/components/Modal/__tests__/ModalHeader.test.tsx:17]: we don't, id removed

File: packages/web-react/src/components/Modal/README.md [hunk 24]
```diff
@@ -165,12 +166,26 @@ accessible name for the dialog, e.g. using the `aria-label` attribute on
...
+
+```jsx
+<Modal id="modal-example">
```
[the remark this answers is not in the document]
Comment (pavelklibani) [reply packages/web-react/src/components/Modal/README.md:174]: updated

File: packages/web-react/src/components/Modal/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (pavelklibani) [reply packages/web-react/src/components/Modal/README.md:171]: updated

File: packages/web-react/src/hooks/useCancelEvent.ts (hunk 5 above)
Comment (pavelklibani) [reply packages/web-react/src/hooks/useCancelEvent.ts:4]: reported here:
https://issues.chromium.org/issues/351867704

File: packages/web-react/src/types/modal.ts (hunk 17 above)
Comment (crishpeen) [reply packages/web-react/src/types/modal.ts:25]: Currently, we set the `Close` in ModalHeader and ModalCloseButton too. Can we set it only once? Then this change would be ok.

Or we can leave it as it was if it was working, but anyway that we set it twice is wrong.

Comment (crishpeen) [reply packages/web-react/src/types/modal.ts:25]: Ha, it is already like this. So LGTM!

File: packages/web/src/js/Modal.ts [hunk 25]
```diff
@@ -1,13 +1,23 @@
...
 import { warning } from './common/utilities';
+import { EVENT_KEY } from './constants';
 import EventHandler from './dom/EventHandler';
 import SelectorEngine from './dom/SelectorEngine';
 import { enableToggleTrigger, ScrollControl, SpiritConfig } from './utils';
 
+// TODO: Remove `handleKeyDown` listener when Chrome fixes the bug,
+// right now Chrome is bugged and sends un-cancelable events, so closing modal based on
+// `cancel` event is not possible in chrome.
+// FireFox and Safari are working fine.
```
Comment (dlouhak) [inline packages/web/src/js/Modal.ts:11]:
```suggestion
// Firefox and Safari are working fine.
```

**Review by dlouhak — APPROVED** [review]

File: packages/web-react/src/hooks/useCancelEvent.ts [hunk 26]
```diff
@@ -1,25 +1,61 @@
-import { useCallback, useEffect, MutableRefObject } from 'react';
+import { useCallback, MutableRefObject } from 'react';
+import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
+
+// TODO: Remove `isOpen` and listeners with `handleKeyDown` when Chrome fixes the bug,
+// right now Chrome is bugged and sends un-cancelable events, so closing modal based on
+// `cancel` event is not possible in chrome.
```
[the remark this answers is not in the document]
Comment (dlouhak) [reply packages/web-react/src/hooks/useCancelEvent.ts:6]: Please, add link to the bug report.
